[Woo POS] Coupons: Fetch coupons from local storage

iamgabrielma opened the pull request:
Closes: #15350   

## Description
This PR starts the work to allow coupons usage into POS by fetching them from the site's local storage. This is a small initial step to reuse what we already have in place, and we might move data fetching entirely to the remote API ([task](https://github.com/woocommerce/woocommerce-ios/issues/15324)) depending on exploration as we progress.

Ideally, the item provider will deal with these, and the POS will have no knowledge of where the data comes from, it will just have access to an array of `POSItem` and these could come over the network, from device, o from somewhere else.

Some caveats from this approach:
- If the merchant hasn't fetched coupons previously in the app, no coupons would be stored.
- If the merchant has disabled coupons in WooCommerce settings, depending the flows they've followed in the app, these could have been sync and flushed results in local storage, so no coupons would be stored.

## Testing information
- In `PointOfSaleItemsController` init, add `loadPointOfSaleCoupons()`
- Using a store with coupons enabled in WC settings, assure that we have coupons in local storage by going to Menu > Coupons
- Run the app, load POS, observe the console showing coupons converted to minimal POSItem objects:
> [
Yosemite.POSItem.coupon(Yosemite.POSCoupon(id: 0EB6B480-27B9-4C36-8C1D-585C9863013A, couponID: 3015)), 
Yosemite.POSItem.coupon(Yosemite.POSCoupon(id: FBC69D84-8554-4DEF-BB8D-D1FDE76F0073, couponID: 1171)), 
Yosemite.POSItem.coupon(Yosemite.POSCoupon(id: 61E54C37-2B2E-494D-9116-058DC5FBDD2B, couponID: 1155)), 
Yosemite.POSItem.coupon(Yosemite.POSCoupon(id: 8E750EEF-6D4C-4BD1-A4C9-2A71D84D87AB, couponID: 1153)), 
...

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

@coderabbitai review

iamgabrielma committed:
Inject storagemanager into item service so we can retrieve coupons from local storage

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Map Coupon to POSItem. Make compile

iamgabrielma committed:
Make storage dependency nullable

We don't necessarily want storage to be nullable in the long term, but kept it as such for now so we don't have to change all the tests

iamgabrielma commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -158,6 +158,14 @@ protocol PointOfSaleItemsControllerProtocol {
+private extension PointOfSaleItemsController {
+    func loadPointOfSaleCoupons() {
+        let posCoupons = itemProvider.providePointOfSaleCoupons()
+        debugPrint(posCoupons)
```
This is just temporary for easier testing 👍

staskus commented on Yosemite/Yosemite/PointOfSale/PointOfSaleItemService.swift:
```diff
@@ -91,6 +99,36 @@ public final class PointOfSaleItemService: PointOfSaleItemServiceProtocol {
 
+    // TODO:
+    // gh-15326 - Return PagedItems<POSItem> instead.
+    public func providePointOfSaleCoupons() -> [POSItem] {
```
I'm a bit conflicted about where to put coupons. On one hand, we describe them as `POSItem` so it makes sense to put them in item service.

However, `providePointOfSaleItems` method does not consider coupons to be part of all those items. 🤔 It suggests to me that it's not entirely related and could be put in `PointOfSaleCouponsService`.

I haven't participated in this item service work during the variable products project and I'm not sure what was the thought of `PointOfSaleItemService` further growth.

On the first look, I more agree to putting it within `PointOfSaleItemService`, maybe we would just need to refine the naming:

`providePointOfSaleItems` -> `providePointOfSaleProductItems`

What are your thoughts on this?

iamgabrielma replied:
This is something that crossed my mind as well, I think at the moment it makes more sense to refine the naming from `providePointOfSaleItems` to `providePointOfSaleProductItems `:

Initially we designed `providePointOfSaleItems` with the idea that we'd make everything an item and all of it will be filtered later on, however with the separation of `providePointOfSaleItems` and `providePointOfSaleVariationItems` we're already implicitly defining what sort of POSItem we'll receive depending what we call, despite knowing that all of them still conform to a common POSItem.

There seems to be also the responsibility of the `PointOfSaleItemsController` to "fetch items, whatever those are depending on the situation" and keep it abstracted in this level, making it more explicit and keeping things separated on the `PointOfSaleItemService`.

This might change of course, but for now I think refining the name is a good change. I'll do this as we move forward 👍

staskus commented on Yosemite/Yosemite/PointOfSale/POSCoupon.swift:
```diff
@@ -0,0 +1,5 @@
+
```
```suggestion
```

staskus reviewed (approved):
Thanks!

I think it's a good first step which enables further development for both coupon fetching and application.

It would be good to sync with other iOS devs if it's intended we expand `PointOfSaleItemService` with coupons.

> This is a small initial step to reuse what we already have in place, and we might move data fetching entirely to the remote API (https://github.com/woocommerce/woocommerce-ios/issues/15324) depending on exploration as we progress.

I think the best would be to combine both local and remote like IPP does. Of course, we'll need to reconsider the API of `PointOfSaleItemServiceProtocol` since we'll want to receive cached items first and then updated ones from the remote. 

> If the merchant hasn't fetched coupons previously in the app, no coupons would be stored.
> If the merchant has disabled coupons in WooCommerce settings, depending the flows they've followed in the app, these could have been sync and flushed results in local storage, so no coupons would be stored.

iamgabrielma committed:
Update Yosemite/Yosemite/PointOfSale/POSCoupon.swift

Co-authored-by: Povilas Staskus <povilas.staskus@automattic.com>

iamgabrielma commented:
@coderabbitai review
